File: cmd/evmsdk/sample-ios/N42Verifier/EvmsdkWrapper.swift
// SPDX-License-Identifier: LGPL-3.0-or-later
//
// Swift wrapper around the gomobile-bound Evmsdk.xcframework. The
// gomobile bind exposes every package-level function in
// cmd/evmsdk/mobile_facade.go as a static method on the `Evmsdk`
// namespace, e.g. Evmsdk.mobileInit(_:).
//
// This wrapper turns those C-style "" / non-empty error strings into
// Swift `throws` and decodes the JSON result blobs into typed structs
// for SwiftUI binding.

import Foundation
import Combine
import Evmsdk     // imported from the gomobile-bound .xcframework

class EvmsdkWrapper: ObservableObject {

    // MARK: - Published state for SwiftUI

    @Published var statusText = "Not initialized"
    @Published var isConnected = false
    @Published var blsKeyHex = ""
    @Published var blocksVerified: Int64 = 0
    @Published var successCount: Int64 = 0
    @Published var failureCount: Int64 = 0
    @Published var successRate: Int64 = 0
    @Published var avgTimeMs: Int64 = 0
    @Published var lastBlockInfo: BlockInfo?
    @Published var verifyLog: [LogEntry] = []
    @Published var producers: [Producer] = []
    @Published var activeProducerURI: String = ""
    @Published var rotationPolicy = RotationPolicy()

    // MARK: - Types

    struct BlockInfo: Identifiable {
        let id = UUID()
        let blockNumber: Int64
        let blockHash: String
        let computedReceiptsRoot: String
        let txCount: Int
        let witnessReadLogLen: Int
        let uncachedBytecodes: Int
        let packetSizeBytes: Int
        let verifyTimeMs: Int64
        let success: Bool
        let signature: String
    }

    struct LogEntry: Identifiable {
        let id = UUID()
        let blockNumber: Int64
        let success: Bool
        let verifyTimeMs: Int64
    }

    struct Producer: Identifiable {
        var id: String { uri }
        let uri: String
        let account: String
        let state: String
        let active: Bool
    }

    struct RotationPolicy {
        var blockTimeMs: Int64 = 1000
        var slowLeaderTimeoutMs: Int64 = 3000
        var rotationCycleMs: Int64 = 7000
        var monitorPeriodMs: Int64 = 500
    }

    enum WrapperError: Error, LocalizedError {
        case notInitialized
        case backend(String)
        case decode(String)

        var errorDescription: String? {
            switch self {
            case .notInitialized: return "Verifier not initialized"
            case .backend(let s): return "Backend error: \(s)"
            case .decode(let s): return "Decode error: \(s)"
            }
        }
    }

    // MARK: - Lifecycle

    func initialize(chainID: Int64 = 4242) {
        let err = EvmsdkMobileInit(chainID)
        if !err.isEmpty {
            DispatchQueue.main.async { self.statusText = "Init failed: \(err)" }
            return
        }
        DispatchQueue.main.async { self.statusText = "Initialized" }
    }

    func setPrivKey(_ hex: String) throws {
        let err = EvmsdkMobileSetPrivKey(hex)
        if !err.isEmpty { throw WrapperError.backend(err) }
        let pub = EvmsdkMobileGetPubkey()
        DispatchQueue.main.async { self.blsKeyHex = pub }
    }

    func setServer(host: String, port: UInt16, account: String) throws {
        let uri = "ws://\(host):\(port)"
        let err = EvmsdkMobileSetServer(uri, account)
        if !err.isEmpty { throw WrapperError.backend(err) }
        refreshProducers()
    }

    /// Adds an additional producer connection. Together with setServer this
    /// is how multi-node mining is configured: one producer per IDC node.
    func addProducer(uri: String, account: String) throws {
        let err = EvmsdkMobileAddProducer(uri, account)
        if !err.isEmpty { throw WrapperError.backend(err) }
        refreshProducers()
    }

    func removeProducer(uri: String) throws {
        let err = EvmsdkMobileRemoveProducer(uri)
        if !err.isEmpty { throw WrapperError.backend(err) }
        refreshProducers()
    }

    func refreshProducers() {
        let json = EvmsdkMobileListProducers()
        let activeURI = EvmsdkMobileGetActiveProducer()
        guard let data = json.data(using: .utf8),
              let list = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return
        }
        let next = list.map { d in
            Producer(
                uri: d["uri"] as? String ?? "",
                account: d["account"] as? String ?? "",
                state: d["state"] as? String ?? "unknown",
                active: d["active"] as? Bool ?? false
            )
        }
        DispatchQueue.main.async {
            self.producers = next
            self.activeProducerURI = activeURI
        }
    }

    /// Pushes the local rotation policy values down to the SDK.
    func applyRotationPolicy() throws {
        let err = EvmsdkMobileSetRotationPolicy(
            rotationPolicy.blockTimeMs,
            rotationPolicy.slowLeaderTimeoutMs,
            rotationPolicy.rotationCycleMs,
            rotationPolicy.monitorPeriodMs
        )
        if !err.isEmpty { throw WrapperError.backend(err) }
    }

    func start() throws {
        let err = EvmsdkMobileStart()
        if !err.isEmpty { throw WrapperError.backend(err) }
        DispatchQueue.main.async {
            self.isConnected = true
            self.statusText = "Connected"
            self.startStatsPoller()
        }
    }

    func stop() {
        _ = EvmsdkMobileStop()
        DispatchQueue.main.async {
            self.isConnected = false
            self.statusText = "Disconnected"
        }
    }

    func free() {
        _ = EvmsdkMobileFree()
    }

    deinit {
        free()
    }

    // MARK: - Stats polling

    private var pollerTimer: Timer?

    private func startStatsPoller() {
        pollerTimer?.invalidate()
        pollerTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.refreshStats()
            self?.refreshLastBlock()
            self?.refreshProducers()
        }
    }

    private func refreshStats() {
        let json = EvmsdkMobileGetStats()
        guard let data = json.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        DispatchQueue.main.async {
            self.blocksVerified = dict["blocks_verified"] as? Int64 ?? 0
            self.successCount   = dict["success_count"]   as? Int64 ?? 0
            self.failureCount   = dict["failure_count"]   as? Int64 ?? 0
            self.successRate    = dict["success_rate"]    as? Int64 ?? 0
            self.avgTimeMs      = dict["avg_time_ms"]     as? Int64 ?? 0
        }
    }

    private func refreshLastBlock() {
        let json = EvmsdkMobileGetLastVerifyInfo()
        guard json != "{}",
              let data = json.data(using: .utf8),
              let dict = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return
        }
        let info = BlockInfo(
            blockNumber: dict["block_number"] as? Int64 ?? 0,
            blockHash:   dict["block_hash"]   as? String ?? "",
            computedReceiptsRoot: dict["computed_receipts_root"] as? String ?? "",
            txCount:     dict["tx_count"]    as? Int ?? 0,
            witnessReadLogLen: dict["witness_read_log_len"] as? Int ?? 0,
            uncachedBytecodes: dict["uncached_bytecodes"]  as? Int ?? 0,
            packetSizeBytes:   dict["packet_size_bytes"]   as? Int ?? 0,
            verifyTimeMs:      dict["verify_time_ms"]      as? Int64 ?? 0,
            success:           dict["success"]              as? Bool ?? false,
            signature:         dict["signature"]            as? String ?? ""
        )
        DispatchQueue.main.async {
            // Only push to log on a new block number
            if self.lastBlockInfo?.blockNumber != info.blockNumber {
                self.verifyLog.insert(LogEntry(
                    blockNumber: info.blockNumber,
                    success: info.success,
                    verifyTimeMs: info.verifyTimeMs
                ), at: 0)
                if self.verifyLog.count > 100 {
                    self.verifyLog = Array(self.verifyLog.prefix(100))
                }
            }
            self.lastBlockInfo = info
        }
    }
}

File: cmd/evmsdk/sample-android/app/src/main/java/io/n42/verifier/EvmsdkWrapper.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
//
// Kotlin wrapper around the gomobile-bound evmsdk.aar. The gomobile
// bind exposes every package-level function in cmd/evmsdk/mobile_facade.go
// as a static method on the `Evmsdk` class, e.g. Evmsdk.mobileInit(4242L).
//
// This wrapper turns "" / non-empty error strings into Kotlin exceptions
// and decodes the JSON result blobs into typed data classes for Compose.

package io.n42.verifier

import evmsdk.Evmsdk   // imported from the gomobile-bound .aar
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject

class EvmsdkWrapper {

    // ─── Published state for Compose ───────────────────────────────────

    private val _status = MutableStateFlow("Not initialized")
    val status: StateFlow<String> = _status

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _blsKeyHex = MutableStateFlow("")
    val blsKeyHex: StateFlow<String> = _blsKeyHex

    private val _stats = MutableStateFlow(Stats())
    val stats: StateFlow<Stats> = _stats

    private val _lastBlock = MutableStateFlow<BlockInfo?>(null)
    val lastBlock: StateFlow<BlockInfo?> = _lastBlock

    private val _verifyLog = MutableStateFlow<List<LogEntry>>(emptyList())
    val verifyLog: StateFlow<List<LogEntry>> = _verifyLog

    private val _producers = MutableStateFlow<List<Producer>>(emptyList())
    val producers: StateFlow<List<Producer>> = _producers

    private val _activeProducerURI = MutableStateFlow("")
    val activeProducerURI: StateFlow<String> = _activeProducerURI

    private val _policy = MutableStateFlow(RotationPolicy())
    val policy: StateFlow<RotationPolicy> = _policy

    // ─── Types ─────────────────────────────────────────────────────────

    data class Stats(
        val blocksVerified: Long = 0,
        val successCount: Long = 0,
        val failureCount: Long = 0,
        val successRate: Long = 0,
        val avgTimeMs: Long = 0,
    )

    data class BlockInfo(
        val blockNumber: Long,
        val blockHash: String,
        val computedReceiptsRoot: String,
        val txCount: Int,
        val witnessReadLogLen: Int,
        val uncachedBytecodes: Int,
        val packetSizeBytes: Int,
        val verifyTimeMs: Long,
        val success: Boolean,
        val signature: String,
    )

    data class LogEntry(
        val blockNumber: Long,
        val success: Boolean,
        val verifyTimeMs: Long,
    )

    data class Producer(
        val uri: String,
        val account: String,
        val state: String,
        val active: Boolean = false,
    )

    data class RotationPolicy(
        val blockTimeMs: Long = 1000,
        val slowLeaderTimeoutMs: Long = 3000,
        val rotationCycleMs: Long = 7000,
        val monitorPeriodMs: Long = 500,
    )

    class WrapperException(message: String) : Exception(message)

    // ─── Lifecycle ─────────────────────────────────────────────────────

    fun initialize(chainID: Long = 4242L) {
        val err = Evmsdk.mobileInit(chainID)
        _status.value = if (err.isEmpty()) "Initialized" else "Init failed: $err"
    }

    fun setPrivKey(hex: String) {
        val err = Evmsdk.mobileSetPrivKey(hex)
        if (err.isNotEmpty()) throw WrapperException(err)
        _blsKeyHex.value = Evmsdk.mobileGetPubkey()
    }

    fun setServer(host: String, port: Int, account: String) {
        val uri = "ws://$host:$port"
        val err = Evmsdk.mobileSetServer(uri, account)
        if (err.isNotEmpty()) throw WrapperException(err)
        refreshProducers()
    }

    /** Adds an additional producer connection for multi-node mining. */
    fun addProducer(uri: String, account: String) {
        val err = Evmsdk.mobileAddProducer(uri, account)
        if (err.isNotEmpty()) throw WrapperException(err)
        refreshProducers()
    }

    fun removeProducer(uri: String) {
        val err = Evmsdk.mobileRemoveProducer(uri)
        if (err.isNotEmpty()) throw WrapperException(err)
        refreshProducers()
    }

    fun refreshProducers() {
        val json = Evmsdk.mobileListProducers()
        val arr = try { org.json.JSONArray(json) } catch (_: Exception) { return }
        val list = mutableListOf<Producer>()
        for (i in 0 until arr.length()) {
            val o = arr.getJSONObject(i)
            list += Producer(
                uri     = o.optString("uri"),
                account = o.optString("account"),
                state   = o.optString("state"),
                active  = o.optBoolean("active"),
            )
        }
        _producers.value = list
        _activeProducerURI.value = Evmsdk.mobileGetActiveProducer()
    }

    fun applyRotationPolicy(p: RotationPolicy) {
        val err = Evmsdk.mobileSetRotationPolicy(
            p.blockTimeMs,
            p.slowLeaderTimeoutMs,
            p.rotationCycleMs,
            p.monitorPeriodMs,
        )
        if (err.isNotEmpty()) throw WrapperException(err)
        _policy.value = p
    }

    fun start() {
        val err = Evmsdk.mobileStart()
        if (err.isNotEmpty()) throw WrapperException(err)
        _isConnected.value = true
        _status.value = "Connected"
        startStatsPoller()
    }

    fun stop() {
        Evmsdk.mobileStop()
        stopStatsPoller()
        _isConnected.value = false
        _status.value = "Disconnected"
    }

    fun free() {
        Evmsdk.mobileFree()
    }

    // ─── Stats polling ─────────────────────────────────────────────────

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var pollerJob: Job? = null

    private fun startStatsPoller() {
        pollerJob?.cancel()
        pollerJob = scope.launch {
            while (isActive) {
                refreshStats()
                refreshLastBlock()
                refreshProducers()
                delay(1000)
            }
        }
    }

    private fun stopStatsPoller() {
        pollerJob?.cancel()
        pollerJob = null
    }

    private fun refreshStats() {
        val json = Evmsdk.mobileGetStats()
        val obj = try { JSONObject(json) } catch (_: Exception) { return }
        _stats.value = Stats(
            blocksVerified = obj.optLong("blocks_verified"),
            successCount   = obj.optLong("success_count"),
            failureCount   = obj.optLong("failure_count"),
            successRate    = obj.optLong("success_rate"),
            avgTimeMs      = obj.optLong("avg_time_ms"),
        )
    }

    private fun refreshLastBlock() {
        val json = Evmsdk.mobileGetLastVerifyInfo()
        if (json == "{}") return
        val obj = try { JSONObject(json) } catch (_: Exception) { return }
        val info = BlockInfo(
            blockNumber          = obj.optLong("block_number"),
            blockHash            = obj.optString("block_hash"),
            computedReceiptsRoot = obj.optString("computed_receipts_root"),
            txCount              = obj.optInt("tx_count"),
            witnessReadLogLen    = obj.optInt("witness_read_log_len"),
            uncachedBytecodes    = obj.optInt("uncached_bytecodes"),
            packetSizeBytes      = obj.optInt("packet_size_bytes"),
            verifyTimeMs         = obj.optLong("verify_time_ms"),
            success              = obj.optBoolean("success"),
            signature            = obj.optString("signature"),
        )
        if (_lastBlock.value?.blockNumber != info.blockNumber) {
            _verifyLog.value = (listOf(
                LogEntry(info.blockNumber, info.success, info.verifyTimeMs)
            ) + _verifyLog.value).take(100)
        }
        _lastBlock.value = info
    }
}
